File: Assets/Bundles/Client/Resources/TS/Controllers/Global/Chat/ChatController.lua
-- Compiled with unity-ts v2.1.0-75
local TS = require("Shared/include/RuntimeLib")
local Reflect = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Reflect
local Controller = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Controller
local GameObjectBridge = require("Shared/TS/GameObjectBridge").GameObjectBridge
local Network = require("Shared/TS/Network").Network
local _UserInput = require("Shared/TS/UserInput/init")
local Keyboard = _UserInput.Keyboard
local Mouse = _UserInput.Mouse
local Bin = require("Shared/TS/Util/Bin").Bin
local CanvasAPI = require("Shared/TS/Util/CanvasAPI").CanvasAPI
local ChatController
do
	ChatController = setmetatable({}, {
		__tostring = function()
			return "ChatController"
		end,
	})
	ChatController.__index = ChatController
	function ChatController.new(...)
		local self = setmetatable({}, ChatController)
		return self:constructor(...) or self
	end
	function ChatController:constructor(localEntityController)
		self.localEntityController = localEntityController
		self.selected = false
		self.selectedBin = Bin.new()
		local refs = GameObject:Find("Chat"):GetComponent("GameObjectReferences")
		self.content = refs:GetValue("UI", "Content")
		self.chatMessagePrefab = refs:GetValue("UI", "ChatMessagePrefab")
		self.inputField = refs:GetValue("UI", "InputField")
		self.content.gameObject:ClearChildren()
	end
	function ChatController:OnStart()
		Network.ServerToClient.ChatMessage.Client:OnServerEvent(function(text)
			self:AddChatMessage(text)
		end)
		local keyboard = Keyboard.new()
		keyboard.KeyDown:ConnectWithPriority(100, function(event)
			if self.selected then
				if event.Key == 2 then
					if self.inputField.text == "" then
						EventSystem.current:ClearSelected()
						return nil
					end
					self:SubmitInputField()
					return nil
				elseif event.Key == 60 then
					EventSystem.current:ClearSelected()
					self.inputField:SetTextWithoutNotify("")
					event:SetCancelled(true)
					return nil
				end
				-- cancel input when using input field
				event:SetCancelled(true)
			elseif event.Key == 2 then
				self.inputField:Select()
			elseif event.Key == 9 then
				self.inputField:SetTextWithoutNotify("/")
				self.inputField.caretPosition = 1
				self.inputField:Select()
			end
		end)
		local mouse = Mouse.new()
		CanvasAPI:OnSelectEvent(self.inputField.gameObject, function()
			self.selected = true
			local _entityInputDisabler = self.localEntityController:GetEntityInput()
			if _entityInputDisabler ~= nil then
				_entityInputDisabler = _entityInputDisabler:AddDisabler()
			end
			local entityInputDisabler = _entityInputDisabler
			if entityInputDisabler ~= nil then
				self.selectedBin:Add(entityInputDisabler)
			end
			local mouseLocker = mouse:AddUnlocker()
			self.selectedBin:Add(function()
				mouse:RemoveUnlocker(mouseLocker)
			end)
		end)
		CanvasAPI:OnDeselectEvent(self.inputField.gameObject, function()
			self.selectedBin:Clean()
			self.selected = false
		end)
	end
	function ChatController:SubmitInputField()
		local text = self.inputField.text
		self:SendChatMessage(text)
		self.inputField:SetTextWithoutNotify("")
		EventSystem.current:ClearSelected()
	end
	function ChatController:SendChatMessage(message)
		Network.ClientToServer.SendChatMessage.Client:FireServer(message)
	end
	function ChatController:AddChatMessage(message)
		TS.try(function()
			local chatMessage = GameObjectBridge:InstantiateIn(self.chatMessagePrefab, self.content.transform)
			local refs = chatMessage:GetComponent("GameObjectReferences")
			local textGui = refs:GetValue("UI", "Text")
			textGui.text = message
		end, function(err)
			print("chat error.")
			print(err)
		end)
	end
	function ChatController:IsChatFocused()
		return self.selected
	end
end
-- (Flamework) ChatController metadata
Reflect.defineMetadata(ChatController, "identifier", "Bundles/Client/Controllers/Global/Chat/ChatController@ChatController")
Reflect.defineMetadata(ChatController, "flamework:parameters", { "Bundles/Client/Controllers/Global/Character/LocalEntityController@LocalEntityController" })
Reflect.defineMetadata(ChatController, "flamework:implements", { "$:flamework@OnStart" })
Reflect.decorate(ChatController, "$:flamework@Controller", Controller, { {} })
return {
	ChatController = ChatController,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Typescript~/src/Client/Controllers/Global/Chat/ChatController.ts
import { Controller, OnStart } from "@easy-games/flamework-core";
import { GameObjectUtil } from "Shared/GameObjectBridge";
import { Network } from "Shared/Network";
import { Keyboard, Mouse } from "Shared/UserInput";
import { Bin } from "Shared/Util/Bin";
import { CanvasAPI } from "Shared/Util/CanvasAPI";
import { SignalPriority } from "Shared/Util/Signal";
import { SetInterval, SetTimeout } from "Shared/Util/Timer";
import { LocalEntityController } from "../Character/LocalEntityController";
import { ChatCommand } from "CoreShared/Commands/ChatCommand";
import { ChatUtil } from "CoreShared/Util/ChatUtil";
import { PlayerController } from "../Player/PlayerController";
import { Game } from "Shared/Game";
import { FriendsCommand } from "CoreShared/Commands/FriendsCommand";
import { encode } from "CoreShared/json";

class ChatMessageElement {
	public canvasGroup: CanvasGroup;
	public shownAt = os.clock();
	public shown = true;
	private hideBin = new Bin();

	constructor(public readonly gameObject: GameObject, public time: number) {
		this.canvasGroup = gameObject.GetComponent<CanvasGroup>();
	}

	public Hide(): void {
		if (!this.shown) return;
		this.shown = false;
		const t = this.canvasGroup.TweenCanvasGroupAlpha(0, 0.2);
		this.hideBin.Add(() => {
			t.Cancel();
		});
	}

	public Show(): void {
		this.shownAt = os.clock();
		if (this.shown) return;
		this.shown = true;
		this.hideBin.Clean();
		this.canvasGroup.alpha = 1;
	}

	public Destroy(): void {
		this.Hide();
		SetTimeout(1, () => {
			Object.Destroy(this.gameObject);
		});
	}
}

@Controller({})
export class ChatController implements OnStart {
	private content: GameObject;
	private chatMessagePrefab: Object;
	private inputField: TMP_InputField;

	private selected = false;
	private selectedBin = new Bin();

	private chatMessageElements: ChatMessageElement[] = [];

	private prevSentMessages: string[] = [];
	private historyIndex = -1;

	private commands = new Map<string, ChatCommand>();

	constructor(private localEntityController: LocalEntityController) {
		const refs = GameObject.Find("Chat").GetComponent<GameObjectReferences>();
		this.content = refs.GetValue("UI", "Content");
		this.chatMessagePrefab = refs.GetValue("UI", "ChatMessagePrefab");
		this.inputField = refs.GetValue("UI", "InputField");
		this.content.gameObject.ClearChildren();

		this.RegisterCommand(new FriendsCommand());
	}

	public RegisterCommand(command: ChatCommand) {
		this.commands.set(command.commandLabel.lower(), command);
		for (let alias of command.aliases) {
			this.commands.set(alias.lower(), command);
		}
	}

	OnStart(): void {
		Network.ServerToClient.ChatMessage.Client.OnServerEvent((text) => {
			this.AddChatMessage(text);
		});

		const keyboard = new Keyboard();
		keyboard.OnKeyDown(
			KeyCode.Return,
			(event) => {
				if (this.selected) {
					if (this.inputField.text === "") {
						EventSystem.current.ClearSelected();
						return;
					}
					this.SubmitInputField();
					event.SetCancelled(true);
				} else {
					this.inputField.Select();
				}
			},
			SignalPriority.HIGH,
		);
		keyboard.OnKeyDown(
			KeyCode.Escape,
			(event) => {
				if (this.selected) {
					if (this.inputField.text === "") {
						EventSystem.current.ClearSelected();
						return;
					}
					this.SubmitInputField();
					event.SetCancelled(true);
				}
			},
			SignalPriority.HIGH,
		);
		keyboard.OnKeyDown(
			KeyCode.Slash,
			(event) => {
				if (!this.selected) {
					this.inputField.SetTextWithoutNotify("/");
					this.inputField.caretPosition = 1;
					this.inputField.Select();
				}
			},
			SignalPriority.HIGH,
		);
		keyboard.OnKeyDown(KeyCode.UpArrow, (event) => {
			print("up.1");
			if (this.IsChatFocused()) {
				print("up.2");
				if (this.historyIndex + 1 < this.prevSentMessages.size()) {
					this.historyIndex++;
					let msg = this.prevSentMessages[this.historyIndex];
					print("up.3 " + msg);
					this.inputField.SetTextWithoutNotify(msg);
					this.inputField.caretPosition = msg.size();
				}
			}
		});
		keyboard.OnKeyDown(KeyCode.DownArrow, (event) => {
			if (this.IsChatFocused()) {
				if (this.historyIndex - 1 >= -1) {
					this.historyIndex--;
					let msg: string;
					if (this.historyIndex === -1) {
						msg = "";
					} else {
						msg = this.prevSentMessages[this.historyIndex];
					}
					this.inputField.SetTextWithoutNotify(msg);
					this.inputField.caretPosition = msg.size();
				}
			}
		});

		// Sink key events when selected:
		keyboard.AnyKeyDown.ConnectWithPriority(SignalPriority.HIGH, (event) => {
			if (this.selected) {
				if (
					event.KeyCode !== KeyCode.Return &&
					event.KeyCode !== KeyCode.Escape &&
					event.KeyCode !== KeyCode.UpArrow &&
					event.KeyCode !== KeyCode.DownArrow
				) {
					event.SetCancelled(true);
				}
			}
		});

		const mouse = new Mouse();
		CanvasAPI.OnSelectEvent(this.inputField.gameObject, () => {
			this.selected = true;
			this.historyIndex = -1;
			const entityInputDisabler = this.localEntityController.GetEntityInput()?.AddDisabler();
			if (entityInputDisabler !== undefined) {
				this.selectedBin.Add(entityInputDisabler);
			}
			const mouseLocker = mouse.AddUnlocker();
			this.selectedBin.Add(() => {
				mouse.RemoveUnlocker(mouseLocker);
			});
			this.CheckIfShouldHide();
		});

		CanvasAPI.OnDeselectEvent(this.inputField.gameObject, () => {
			this.selectedBin.Clean();
			this.selected = false;
			this.CheckIfShouldHide();
		});

		SetInterval(0.5, () => {
			this.CheckIfShouldHide();
		});
	}

	private CheckIfShouldHide(): void {
		if (this.IsChatFocused()) {
			for (const element of this.chatMessageElements) {
				element.Show();
			}
		} else {
			for (const element of this.chatMessageElements) {
				if (os.clock() - element.time > 5) {
					element.Hide();
				}
			}
		}
	}

	public ShowAllChatMessages(): void {
		for (const element of this.chatMessageElements) {
			element.Show();
		}
	}

	public SubmitInputField(): void {
		let text = this.inputField.text;
		this.SendChatMessage(text);
		this.inputField.SetTextWithoutNotify("");
		EventSystem.current.ClearSelected();
	}

	public SendChatMessage(message: string): void {
		this.prevSentMessages.unshift(message);
		if (this.prevSentMessages.size() > 50) {
			this.prevSentMessages.pop();
		}

		const commandData = ChatUtil.ParseCommandData(message);

		print(`SendChatMessage() commandData: ${encode(commandData)}`);

		let sendChatToServer = true;

		if (commandData) {
			const command = this.commands.get(commandData.label);
			if (command) {
				command.Execute(Game.LocalPlayer, commandData.args);
				sendChatToServer = false;
			}
		}

		if (sendChatToServer) {
			Network.ClientToServer.SendChatMessage.Client.FireServer(message);
		}
	}

	public AddChatMessage(message: string): void {
		try {
			const chatMessage = GameObjectUtil.InstantiateIn(this.chatMessagePrefab, this.content.transform);
			const refs = chatMessage.GetComponent<GameObjectReferences>();
			const textGui = refs.GetValue<TextMeshProUGUI>("UI", "Text");

			textGui.text = message;

			const element = new ChatMessageElement(chatMessage, os.clock());
			this.chatMessageElements.push(element);
		} catch (err) {
			Debug.LogError("chat error:");
			Debug.LogError(err);
		}
	}

	public IsChatFocused(): boolean {
		return this.selected;
	}
}
